### PR TITLE
Terraform: GROK-12085 s3 datagrok_pubic backup

### PR DESCRIPTION
**Pull request by ksumag**
creating iam backup role, policy, selection and plan

### CODE REVIEW

File: aws/s3.tf [hunk 1]
```diff
@@ -123,3 +123,58 @@ module "s3_bucket" {
...
+        }
+        Action = "sts:AssumeRole"
+      }
+    ]
+  })
+}
+
+# Attach an IAM policy to the backup role that allows  performing AWS Backup jobs
+resource "aws_iam_role_policy_attachment" "backup_policy_attachment" {
+  policy_arn = "arn:aws:iam::aws:policy/service-role/AWSBackupServiceRolePolicyForBackup"
```
Comment (dnillovna) [inline aws/s3.tf:150]: Is it secure enough?

File: aws/s3.tf [hunk 2]
```diff
@@ -123,3 +123,58 @@ module "s3_bucket" {
...
+  role       = aws_iam_role.backup_role.name
+}
+resource "aws_backup_plan" "datagrok_public_s3_backup_plan" {
+  name        = "datagrok_public_s3_backup_plan"
+  description = "datagrok_public s3 bucket backup plan"
```
Comment (dnillovna) [inline aws/s3.tf:155]: Use variables

File: aws/s3.tf (hunk 1 above)
Comment (ksumag) [reply aws/s3.tf]: why not? arn:aws:iam::aws:policy/service-role/AWSBackupServiceRolePolicyForBackup is all known policy

Comment (ksumag) [reply aws/s3.tf]: this policy is needed for Iam role can do backups. is it something wrong?

File: aws/s3.tf [hunk 3]
```diff
@@ -123,3 +123,94 @@ module "s3_bucket" {
 
   tags = local.tags
 }
+# aws S3 backup //////////////////////
+resource "aws_backup_vault" "datagrok_public_vault" {
+  name = var.s3_backup_vault_name
+}
+
+resource "aws_iam_role" "backup_role" {
+  name                = "s3-backup-role"
```
Comment (dnillovna) [inline aws/s3.tf:132]: The name should be with the same prefixes as other resources to determine that it was created for our resources
Like for the logging bucket https://github.com/datagrok-ai/tf-module-datagrok-core/pull/26/files#diff-bfa811139a4cedd8c743561172e24c14f800b2fd054b137f984598b3d4195706R10

File: aws/s3.tf [hunk 4]
```diff
@@ -123,3 +123,94 @@ module "s3_bucket" {
...
+          Service = "backup.amazonaws.com"
+        }
+        Action = "sts:AssumeRole"
+      }
+    ]
+  })
+}
+
+resource "aws_iam_policy" "s3_backup" {
+  name        = "backup-s3"
```
Comment (dnillovna) [inline aws/s3.tf:148]: The name should be with the same prefixes as other resources to determine that it was created for our resources
Like for the logging bucket https://github.com/datagrok-ai/tf-module-datagrok-core/pull/26/files#diff-bfa811139a4cedd8c743561172e24c14f800b2fd054b137f984598b3d4195706R10

File: aws/s3.tf [hunk 5]
```diff
@@ -123,3 +123,94 @@ module "s3_bucket" {
...
+  description = "policy for backup s3 datagrok-public bucket"
```
Comment (dnillovna) [inline aws/s3.tf:149]: datagrok-public is a hardcoded name. It is a name for our bucket and won't be the same for clients.
You should use `s3_name` here instead

File: aws/s3.tf [hunk 6]
```diff
@@ -123,3 +123,94 @@ module "s3_bucket" {
...
+resource "aws_iam_role_policy_attachment" "backup_policy_attachment" {
+  policy_arn = aws_iam_policy.s3_backup.arn
+  role       = aws_iam_role.backup_role.name
+}
+resource "aws_iam_role_policy_attachment" "backup_service_role_policy" {
+  policy_arn = "arn:aws:iam::aws:policy/AWSBackupServiceRolePolicyForS3Backup"
+  role       = aws_iam_role.backup_role.name
+}
+resource "aws_backup_plan" "datagrok_public_s3_backup_plan" {
+  name = var.s3_backup_plan_name
```
Comment (dnillovna) [inline aws/s3.tf:191]: We already have a lot of variables; why would we need another one?
Use naming prefixes instead, like for logging bucket https://github.com/datagrok-ai/tf-module-datagrok-core/pull/26/files#diff-bfa811139a4cedd8c743561172e24c14f800b2fd054b137f984598b3d4195706R10

File: aws/s3.tf [hunk 7]
```diff
@@ -123,3 +123,94 @@ module "s3_bucket" {
...
+
+  rule {
+    rule_name         = "Daily-S3-backups-rule"
+    target_vault_name = aws_backup_vault.datagrok_public_vault.name
+    schedule          = var.s3_backup_schedule
+
+    lifecycle {
+      delete_after = "14"
```
Comment (dnillovna) [inline aws/s3.tf:199]: It should be configurable

File: aws/s3.tf [hunk 8]
```diff
@@ -123,3 +123,94 @@ module "s3_bucket" {
...
+# Attach an IAM policy to the backup role that allows  performing AWS Backup jobs
+resource "aws_iam_role_policy_attachment" "backup_policy_attachment" {
+  policy_arn = aws_iam_policy.s3_backup.arn
+  role       = aws_iam_role.backup_role.name
+}
+resource "aws_iam_role_policy_attachment" "backup_service_role_policy" {
+  policy_arn = "arn:aws:iam::aws:policy/AWSBackupServiceRolePolicyForS3Backup"
+  role       = aws_iam_role.backup_role.name
+}
+resource "aws_backup_plan" "datagrok_public_s3_backup_plan" {
```
Comment (dnillovna) [inline aws/s3.tf:190]: The resource name will be the same for us and for clients. It is not right to add our bucket name to the name of the resource. The name should be general, like for other resources. For example, `s3_backup_plan`

File: aws/s3.tf [hunk 9]
```diff
@@ -123,3 +123,94 @@ module "s3_bucket" {
 
   tags = local.tags
 }
+# aws S3 backup //////////////////////
+resource "aws_backup_vault" "datagrok_public_vault" {
```
Comment (dnillovna) [inline aws/s3.tf:127]: The resource name will be the same for us and for clients. It is not right to add our bucket name to the name of the resource. The name should be general, like for other resources. For example, `s3_backup_plan`

File: aws/s3.tf [hunk 10]
```diff
@@ -123,3 +123,94 @@ module "s3_bucket" {
...
+  name = var.s3_backup_vault_name
```
Comment (dnillovna) [inline aws/s3.tf:128]: We already have a lot of variables; why would we need another one?
Use naming prefixes instead, like for logging bucket https://github.com/datagrok-ai/tf-module-datagrok-core/pull/26/files#diff-bfa811139a4cedd8c743561172e24c14f800b2fd054b137f984598b3d4195706R10

**Review by dnillovna — CHANGES_REQUESTED** [review]

File: aws/s3.tf [hunk 11]
```diff
@@ -123,3 +123,94 @@ module "s3_bucket" {
 
   tags = local.tags
 }
+# aws S3 backup //////////////////////
+resource "aws_backup_vault" "s3_backup_vault" {
+  name = "s3-backup-vault-${var.name}-${var.environment}"
```
Comment (dnillovna) [inline aws/s3.tf:128]: A prefix is better than a suffix in this case.

Comment (ksumag) [reply aws/s3.tf]: ok! rewrited!

File: aws/s3.tf [hunk 12]
```diff
@@ -123,3 +123,94 @@ module "s3_bucket" {
...
+resource "aws_backup_plan" "s3_backup_plan" {
+  name = "${var.name}-${var.environment}-s3-backup-plan"
+
+  rule {
+    rule_name         = "Daily-S3-backups-rule"
+    target_vault_name = aws_backup_vault.s3_backup_vault.name
+    schedule          = var.s3_backup_schedule
+
+    lifecycle {
+      delete_after = "${var.s3_backup_lifecycle}"
```
Comment (dnillovna) [inline aws/s3.tf:199]: there is no need in interpolation if it is just one variable :)

Comment (dnillovna) [reply aws/s3.tf:199]: Do you use interpolation because the variable is the type number?

File: aws/variables.tf [hunk 13]
```diff
@@ -200,6 +200,13 @@ variable "s3_name" {
   description = "The name of S3 bucket for Datagrok. If it is not specified, the name along with the environment will be used."
 }
 
+variable "s3_backup_lifecycle" {
+  type        = number
+  default     = 14
+  nullable    = true
+  description = "Describes how many days save s3 backup snapshot."
```
Comment (dnillovna) [inline aws/variables.tf:207]: save is not the right word here. I think you meant `store`

File: aws/s3.tf (hunk 12 above)
Comment (ksumag) [reply aws/s3.tf:199]: Yes, I do

Comment (dnillovna) [reply aws/s3.tf:199]: Well, fine, no objections

File: aws/variables.tf [hunk 14]
```diff
@@ -627,6 +634,13 @@ variable "docker_grok_connect_tag" {
   description = "Tag from Docker Registry for Datagrok Grok Connect Image"
 }
 
+variable "s3_backup_schedule" {
+  type        = string
+  default     = "cron(0 3 * * ? *)"
+  nullable    = false
+  description = "Schedule for backup aws s3 bucket. By default is everyday 3 AM"
```
[the remark this answers is not in the document]
Comment (ksumag) [reply aws/variables.tf]: only split "every day"

File: aws/s3.tf [hunk 15]
```diff
@@ -123,3 +123,100 @@ module "s3_bucket" {
 
   tags = local.tags
 }
+# aws S3 backup //////////////////////
+resource "aws_backup_vault" "s3_backup_vault" {
+  name = "${var.name}-${var.environment}-s3-backup-vault"
+  kms_key_arn = aws_kms_key.s3_backup_kms_key.arn
```
Comment (dnillovna) [inline aws/s3.tf:129]: Use the same `kms_key_arn` as in the rest of the code

File: aws/README.md [hunk 16]
```diff
@@ -1,20 +1,242 @@
...
-  environment         = "example"
-  domain_name         = "datagrok.example"
-  docker_hub_credentials = {
-    create_secret = true
-    user          = "exampleUser"
-    password      = "examplePassword"
-  }
-}
-```
+## Requirements
```
Comment (dnillovna) [inline aws/README.md:1]: Remove the duplicates and return Usage block back

**Pull request merged**